### ISSUE TITLE
Program codegen: Python

### ISSUE THREAD

**Issue by pgavlin** [pulumi/pulumi#4332]
- [ ] Resource variables
	- [x] Basic instantiations
	- [x] Ranged instantiations
		- [x] Eventually-typed ranges (i.e. ranges that are promises or outputs)
	- [ ] Resource options
- [ ] Config variables
	- [ ] Typed config
	- [x] Required config
	- [x] Optional config
- [x] Local variables
- [x] Output variables
- [ ] Component declarations
- [ ] Component variables
- [ ] Expressions
	- [x] Anonymous function expressions
	- [x] Binary operator expressions
	- [x] Conditional expressions
	- [x] For expressions
	- [ ] Function call expressions
		- [x] Apply intrinsic
		- [ ] Interpolate intrinsic (where necessary)
		- [x] `entries`
		- [x] `fileArchive`, `fileAsset`
		- [x] `invoke`
		- [ ] `mimeType`
		- [x] `range`
		- [ ] `readDir`
		- [x] `toJSON`
		- [ ] etc.
	- [x] Index expressions
	- [x] Literal value expressions
		- [x] Multi-line string literals
	- [x] Object construction expressions
	- [x] Relative traversal expressions
	- [x] Scope traversal expressions
	- [x] Splat expressions
	- [x] Template expressions
	- [ ] Template join expressions
	- [x] Tuple construction expressions
	- [x] Unary operator expressions
- [ ] Applies
	- [x] Proxied apply lowering
	- [ ] Interpolate lowering
- [ ] Trivia
	- [ ] Comments
	- [ ] Whitespace

**Comment by pgavlin** [comment]
Closing this for now. We can open follow-up items as necessary.

**Comment by iridian-ks** [comment]
@pgavlin I've been trying to follow the tf2pulumi python backend progress. Once this is done I'd like to try porting some terraform state into pulumi. Are there any pre-releases that I can try? I've got some state with hundreds of resources so maybe I could help find issues.

**Comment by ablekh** [comment]
@pgavlin What is the point of closing an issue that has not yet been resolved or superseded by another one? It would be great if you could clarify the status of this effort.

**Comment by leezen** [comment]
@ablekh Just clarifying the status here. An initial version of Python program code generation is in pulumi/pulumi and any issues with that implementation are tracked as separate issues, e.g. https://github.com/pulumi/pulumi/issues/4815 and https://github.com/pulumi/pulumi/issues/4314. If you have a specific issue you've run into w.r.t. python program code generation please open an issue in this repo.

**Comment by leezen** [comment]
If you're looking for tf2pulumi Python support, this is available in https://github.com/pulumi/tf2pulumi since 0.8.0

**Comment by ablekh** [comment]
@leezen Thank you very much for your prompt feedback. I understand now. The reason for my confusion was frequent mention of TypeScript as (presumably, the only) target language for Terraform-to-Pulumi conversion in `tf2pulumi` repo description (except for several places in Usage section that mention other targets). It is most likely too small of a problem to warrant creating a separate issue, so I hope that it will be taken care of by Pulumi team based on this comment.
